src/libtls/tls_server.c

```diff
@@ -317,16 +444,77 @@ static status_t process_client_hello(private_tls_server_t *this,
 			this->alert->add(this->alert, TLS_FATAL, TLS_HANDSHAKE_FAILURE);
 			return NEED_MORE;
 		}
-		rng = lib->crypto->create_rng(lib->crypto, RNG_STRONG);
-		if (!rng || !rng->allocate_bytes(rng, SESSION_ID_SIZE, &this->session))
+		if (this->tls->get_version_max(this->tls) < TLS_1_3)
 		{
-			DBG1(DBG_TLS, "generating TLS session identifier failed, skipped");
+			rng = lib->crypto->create_rng(lib->crypto, RNG_STRONG);
+			if (!rng ||
+				!rng->allocate_bytes(rng, SESSION_ID_SIZE, &this->session))
+			{
+				DBG1(DBG_TLS, "generating TLS session identifier failed, skipped");
+			}
+			DESTROY_IF(rng);
+		}
+		else
+		{
+			this->session = chunk_clone(session);
 		}
-		DESTROY_IF(rng);
 		DBG1(DBG_TLS, "negotiated %N using suite %N",
 			 tls_version_names, this->tls->get_version_max(this->tls),
 			 tls_cipher_suite_names, this->suite);
 	}
+
+	if (this->tls->get_version_max(this->tls) >= TLS_1_3)
+	{
+		diffie_hellman_group_t group;
+		tls_named_group_t curve;
+		enumerator_t *enumerator;
+		chunk_t shared_secret = chunk_empty;
+
+		enumerator = this->crypto->create_ec_enumerator(this->crypto);
+		while (enumerator->enumerate(enumerator, &group, &curve))
+		{
+			if (peer_supports_curve(this, curve) && peer_supports_curve(this, key_type))
+			{
+				if (!this->dh)
+				{
+					this->dh = lib->crypto->create_dh(lib->crypto, group);
+					break;
+				}
+			}
+		}
+		enumerator->destroy(enumerator);
+		if (!this->dh)
+		{ /* simply fail since HelloRetryRequest is not currently implemented */
```

Comment:
Either put the comment on a separate line after the `{`, or indent it with a tab so it aligns with the code in this block.